ftprime/argrecorder.py

```diff
@@ -208,18 +248,60 @@ def add_individual(self, input_id, time,
         :param float time: The time of birth of the individual.
         :param flags int: Any msprime flags to record (probably not needed).
         :param population int: The population ID of birth of the indivdiual
-            (may be omitted).  
+            (may be omitted).
         '''
-        if input_id not in self.node_ids:
-            self.node_ids[input_id] = self.nodes.num_rows
-            self.nodes.add_row(flags=flags, population=population,
-                               time=time)
-            self.max_time = max(self.max_time, time)
-        else:
-            # nothing bad happens if we try to add an individual more than once,
-            # but this is helpful for debugging
-            raise ValueError("Attempted to add " + str(input_id) +
-                             ", who already exits, as a new individual.")
+        self.add_individuals((input_id, ), (time, ), (flags, ), (population, ))
+
+    def add_records(self, lefts, rights, parents, childrens):
+        '''
+        Add multiple records corresponding to a reproduction event in which
+        children (a tuple of IDs) inherit from parent (a single ID) on the
+        interval [left,right).
+
+        :param iterable of float left: The left endpoint of the chromosomal
+            segment inherited.
+        :param iterable of float right: The right endpoint of the chromosomal
+            segment inherited.
+        :param iterable of int parent: The input ID of the parent.
+        :param iterable of tuple childrens: An iterable of input IDs of the
+            children.
+        '''
+        out_parents = (self.node_ids[parent]
+                       for parent in self._valid_parents(parents))
+        out_childrens = (tuple(self.node_ids[u] for u in children)
+                         for children in childrens)
+        edges = zip(lefts, rights, out_parents, out_childrens)
+        edges_flat = ((left, right, parent, child)
+                      for left, right, parent, children in edges
+                      for child in children)
+        et = np.fromiter(edges_flat, dtype=edge_dt)
+        self.__edges.append_columns(parent=et['parent'],
+                                  child=et['child'],
+                                  left=et['left'],
+                                  right=et['right'])
+
+    def _valid_parents(self, parents):
+        """
+        Check that all ``parents`` are valid.
+        """
+        for p in parents:
+            if p not in self.node_ids:
+                raise ValueError("Parent " + str(p) +
+                                 "'s birth time has not been recorded with " +
+                                 ".add_individual().")
+            yield p
+
+    def _valid_input_ids(self, input_ids):
+        """
+        Check that all ``input_ids`` are valid.
+        """
+        # nothing bad happens if we try to add an individual more than
```

Comment:
this note is no longer true - now `add_individual()` *will* add the same individual more than once to the output node table, which is bad.